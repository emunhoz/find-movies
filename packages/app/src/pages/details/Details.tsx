import React, { useEffect, useState } from 'react'
import { Layout, Button, LabelWithLogo, Loading } from '@monorepo/ui-components'
import { Logo, ArrowWhite } from '@commons/images'
import { findMoviesById } from '../../services/movies'
import { useParams } from 'react-router-dom'
import { IconHeartWhite, IconHeartFull } from '@commons/images'
import Icons from '../../helpers/icons'
import * as S from './Details.style'

interface CustomParams {
  id: string
}

function Details() {
  const [details, setDetails] = useState<any>({})
  const [favourite, setFavourite] = useState<boolean>(false)
  let params = useParams()

  const imdbId = params as CustomParams

  useEffect(() => {
    getDetailData(imdbId.id)
  }, [imdbId])

  async function getDetailData(imdbId: string) {
    try {
      const { data } = await findMoviesById(imdbId)
      setDetails(data)
    } catch (error) {
      console.error(error)
    }
  }

  if (!Object.keys(details).length) {
    return <Loading />
  }

  const actors = details?.actors.split(',')
  const genres = details?.genre?.split(',')
  const directors = details?.director?.split(',')

  return (
    <Layout>
      <S.Container>
        <S.Logo src={Logo} alt="logo" width="112" height="32" />
        <S.LinkWrapper to="/"><Button aria-label="Back to home" icon={ArrowWhite} ghost size="big"></Button></S.LinkWrapper>
        <S.InfoWrapper>
          <S.Details>
            <S.Time>
              <S.LabelWithDot>{details.runtime}</S.LabelWithDot>
              <S.LabelWithDot>{details.year}</S.LabelWithDot>
              <S.Label>{details.rated}</S.Label>
            </S.Time>
            <S.MovieTitle>{details.title}</S.MovieTitle>
            <S.LabelWrapper>
              {details?.ratings?.map((item: { source: any | undefined; value: string | undefined }, key: string | number | null | undefined) => (
                <LabelWithLogo key={key} backgroundLogo={Icons[item.source].background} value={item.value} logo={Icons[item.source].img} />
              ))}
              <Button active={favourite} onClick={() => setFavourite(!favourite)} icon={favourite ? IconHeartFull : IconHeartWhite}>
                {favourite ? 'Added' : 'Add to favourites'}
              </Button>
            </S.LabelWrapper>
            <S.Infos>
              <S.BlockDescription>
                <S.SubTitle>Plot</S.SubTitle>
                <S.Text>{details.plot}</S.Text>
              </S.BlockDescription>
              <S.SubBlock>
                <S.BlockDescription>
                  <S.SubTitle>Cast</S.SubTitle>
                  <S.List>
                    {actors.map((actor: string, key: number) => (
                      <S.ListItem key={key}>{actor}</S.ListItem>
                    ))}
                  </S.List>
                </S.BlockDescription>
                <S.BlockDescription>
                  <S.SubTitle>Genre</S.SubTitle>
                  <S.List>
                    {genres.map((genre: string, key: number) => (
                      <S.ListItem key={key}>{genre}</S.ListItem>
                    ))}
                  </S.List>
                </S.BlockDescription>
                <S.BlockDescription>
                  <S.SubTitle>Director</S.SubTitle>
                  <S.List>
                    {directors.map((director: string, key: number) => (
                      <S.ListItem key={key}>{director}</S.ListItem>
                    ))}
                  </S.List>
                </S.BlockDescription>
              </S.SubBlock>
            </S.Infos>
          </S.Details>
          <S.PosterImage>
            <S.Poster loading="lazy" width="484" height="721" src={details.poster} alt={details.title} />
          </S.PosterImage>
        </S.InfoWrapper>
      </S.Container>
    </Layout>
  )
}

export default Details
