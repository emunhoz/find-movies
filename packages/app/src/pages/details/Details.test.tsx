import React from 'react'
import { render, screen } from '../../../test-utils'
import { MemoryRouter } from 'react-router-dom'
import Details from './Details'

describe('[Page]: Details', () => {
  test('render page', () => {
    const { asFragment } = render(<MemoryRouter><Details /></MemoryRouter>)

    expect(asFragment()).toMatchSnapshot()
  })

  test('shows loading when mount the page', () => {
    render(<MemoryRouter><Details /></MemoryRouter>)
    expect(screen.getByTestId('loading')).toBeTruthy()
  })
})
