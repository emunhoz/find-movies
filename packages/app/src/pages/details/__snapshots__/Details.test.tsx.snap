// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Page]: Details render page 1`] = `
<DocumentFragment>
  <div
    class="sc-fodVxV gHPDdw"
  >
    <div
      class="spinner"
      data-testid="loading"
    />
  </div>
</DocumentFragment>
`;
